text/0000-monorepo-support.md

```diff
@@ -0,0 +1,81 @@
+- Feature Name: monorepo_support
+- Start Date: 2020-04-23
+- RFC PR: (leave this empty)
+- Volta Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Improve the ergonomics of working with Volta in a monorepo environment.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Volta treats the first `package.json` file it finds in the directory tree as the root of the current project and doesn't look any further for more context. This means that users in a monorepo environment are forced to duplicate their Volta settings in each subproject. Additionally, if they ever want to update their pinned `node` or `yarn` versions, they need to update each subproject individually, negating many of the reasons for using a monorepo in the first place. Volta should provide more ergonomic solutions so users in a monorepo can manage their tool versions in a single place without having to duplicate the settings in each subproject.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+The changes proposed by this RFC are primarily advanced features required by users who _manage_ monorepo projects. Users who are purely consumers should be able to continue using Volta without needing to consider the details of how Volta determines the appropriate context. For the advanced users, we will need to document:
+
+- Volta's model for having multiple project roots, using the `"extends"` key in `package.json`
+- The logic Volta uses to combine platform settings from multiple locations into a single context (including the ability to disable inheriting with the `false` value).
+- The additional impact of having multiple project roots, i.e. `hooks` and local binary resolution.
+- The behavior of `volta pin` in subprojects vs the root project.
+
+Ideally, we should document all of these with specific examples to show how the different pieces interact. We could potentially even create an `examples` repo (or directory within the main repo) for users to reference.
+
+# Details
+[details]: #details
+
+## Multiple Roots
+
+The core change to Volta will be expanding our `Project` model to support a _sequence_ of project roots, instead of a single project root. We will support a sequence by adding the ability to put an `"extends"` entry inside the `"volta"` section of `package.json`. The value of that entry will be the path to another json file, which we will parse just like the original `package.json`. The directory that new file is in will be the next value in the sequence of project roots. Then, if that new file also has an `"extends"` entry, we will continue the same process, building up effectively a linked list of project roots to use for all of our project information.
```

Comment:
This prose reads well, but I am struggling to fully understand what the actual contents of `volta` property in the `package.json`'s will be. Can you give a basic folder structure diagram (ala from the `tree` or `exa --tree` commands) and then specify what the contents would be for each `package.json`?

Something like:

---

For example, given the following folder structure:

```
.
├── package.json
└── packages
   ├── bar
   │  └── package.json
   └── foo
      └── package.json
```

The `package.json` files would have the following contents:

* `package.json`
```json
{
  "volta": {
    "node": "12.5.9",
    "yarn": "1.22.0"
  }
}
```

* `packages/bar/package.json`

```json
{
  "volta": {
    "extends": "../../package.json"
  }
}
```


* `packages/foo/package.json`

```json
{
  "volta": {
    "extends": "../../package.json"
  }
}
```